### PR TITLE
Escaped New Line for React Fathead

### LINKED ISSUE #473: JavaScript: Create New ReactJS Fathead
# ReactJS Fathead

Contribute to DuckDuckGo!

screen shot 2016-10-21 at 16 36 35
### What you will need to know?

Either **JavaScript**, Python, Ruby, or Perl for Fatheads. Bash is also useful. You will download the reference source and create an **output.txt** file containing the parsed documentation. Read more about Fatheads here: 
https://docs.duckduckhack.com/resources/fathead-overview.html
### The source :rocket:

https://facebook.github.io/react/docs/top-level-api.html
https://facebook.github.io/react/docs/component-api.html
### The goal

Help _[improve JavaScript language search results](https://forum.duckduckhack.com/t/javascript-search-overview)_ on DuckDuckGo as part of our Programming Mission! We're trying to reach 100% Instant Answer (IA) coverage for JavaScript-related searches by creating new Instant Answers or improving existing ones.

Code to some relevant examples:
- [jQuery Reference](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/jquery)
- [NodeJS Reference](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/node_js)
- [MDN JS Reference](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/mdnjs)

**Fathead Example (jQuery Reference)**

screen shot 2016-10-21 at 17 06 36
### The steps
- [x] Claim this issue (Comment Below)
- [x] Please review our [Contributing Guide](https://github.com/duckduckgo/zeroclickinfo-fathead/blob/master/CONTRIBUTING.md)
- [x] Set up codio environment (See: https://docs.duckduckhack.com/welcome/setup-dev-environment.html) and fork this repository
- [x] Create the Fathead
- [x] Create a PR
- [x] Ping @bfmags for a review
### Questions?
- [Documentation](https://docs.duckduckhack.com/)
- [The Forum](https://forum.duckduckhack.com/c/programming/javascript)
- Join [DuckDuckHack Slack](https://quackslack.herokuapp.com/) 
- Ping one of the [Community Leaders](https://duck.co/help/community/community-leaders) or @bfmags

---

Instant Answer Page: https://duck.co/ia/view/react_api_docs

_Thanks to @pjhampton for the template_

### PR DESCRIPTION
**Pull request by dip-kush**
## Type of Change

- [x] New Instant Answer
- [x] Improvement
  - [ ] Bug fix
  - [ ] Enhancement
- [ ] Non–Instant Answer
  - [ ] Other (Role, Template, Test, Documentation, etc.)
## Description of new Instant Answer, or changes

Fixes #473 
## People to notify

@MariagraziaAlastra 
@bfmags 

---

IA Page: https://duck.co/ia/view/reactjs

Forum Topic: [Create IA: ReactJS Fathead](https://forum.duckduckhack.com/t/create-ia-reactjs-fathead/999)

Fixes #473

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by bfmags** [comment]
@moollaza can you check it and see if it's OK now?

**Comment by moollaza** [comment]
@bfmags looks like we have `\\n` in the abstract now instead of `\n`

Where there is a literal `\n` in the abstract, we will put in a proper newline character. Where there is a `\\n`, you will see the literal string `\n` displayed.

So `\n` in the output.txt is like putting a `<br>` -- does that clarify things?

**Comment by bfmags** [comment]
reverted the changes :: back to the beginning :)

**Comment by dip-kush** [comment]
@bfmags @moollaza Any Suggestion what could be done here  ?

**Comment by moollaza** [comment]
@dip-kush this is looking correct now, however there's one more issues -- the directory name, and the Fathead ID need to match. Can you rename the dir from `react_js` to `reactjs`?

We use the ID to lookup the code in the repo which is why they need to match 👍

**Comment by moollaza** [comment]
@dip-kush @bfmags I went ahead and changed the directory name locally so I could get this on Beta: https://beta.duckduckgo.com/?q=React.component&t=h_&ia=about&iax=1

The newlines are all correct now 😃 

Once the directory is changed this can be merged 👍

**Comment by moollaza** [comment]
@dip-kush actually looking at this a little closer I see that we're pulling in the HTML from the code block in the documentation, and we're pulling in `<a>` tags as well, causing problems.

We should be stripping away all HTML from inside the code block we copy, so we're just left with the plaintext code. We have syntax highlighting that we'll use on our end to make it look good.

We currently can't handle `<a>` tags in the abstract very well, so we should remove those as well. 

To recap:

1. Change the dir from `react_js` to `reactjs`
2. Remove all HTML from within the code blocks. The ouput.txt should contain `<pre><code>{{plaintext code snippet}}</code></pre>
3. Remove all `<a>` from the description text we scrape.

**Comment by dip-kush** [comment]
I will make the recommended changes. Thanks for the comments.

**Comment by dip-kush** [comment]
@moollaza I have made the required changes. 
1. Change the dir from react_js to reactjs
2.Remove all HTML from within the code blocks. The ouput.txt should contain `
{{plaintext code snippet}}
3.Remove all <a> from the description text we scrape.

File: lib/fathead/reactjs/parse.py [hunk 1]
```diff
@@ -95,8 +95,15 @@ def parse_example(self, api):
             if tag.name == 'div':
                 code = str(tag.find('pre'))
                 if code != 'None':
-                    example_code = code
-        return example_code.strip('\n')
+                    example_code = tag.text.strip()
+        example_code+='\n'
+        return example_code.replace('\n','\\n')
```
Comment (moollaza) [inline lib/fathead/reactjs/parse.py]: @dip-kush looks like we also have instances of multiple newlines (e.g. `\n\n`) We should normalize those to a single newline

File: lib/fathead/reactjs/parse.py [hunk 2]
```diff
@@ -148,6 +157,7 @@ def create_file(self):
             example = data_element['example']
             content = data_element['content']
             if example:
+                example = '<pre><code>%s</code></pre>' % example
                 abstract = '<section class="prog__container">{}<br>{}</section>'.format(content, example)
```
Comment (moollaza) [inline lib/fathead/reactjs/parse.py]: There should be no `<br>` tags -- please remove that.

**Comment by dip-kush** [comment]
@moollaza Made the above requested changes.

**Comment by moollaza** [comment]
@dip-kush thanks for the updates! This is live on Beta:

https://beta.duckduckgo.com/?q=React.Component&ia=about&iax=1

---

I noticed one odd result: https://beta.duckduckgo.com/?q=clipboard+events&ia=about -- it looks like we're missing some data there?

Same here: https://beta.duckduckgo.com/?q=animation+events&ia=about

-----

For https://beta.duckduckgo.com/?q=React.PropTypes&ia=about&iax=1

We should be replacing the `<h4><code>` with `<span class="prog__sub">` -- this is the proper syntax for sub-headings 👍  

---

This is good enough to be merged though. Nice work @dip-kush 💯

**Comment by moollaza** [comment]
@dip-kush after looking a little closer, I see we seem to be showing the wrong output for several of the `React.PropTypes.*` methods.

https://beta.duckduckgo.com/?q=React.PropTypes.arrayOf()&ia=about&iax=1

If you click the "More at" link you'll see they have a code snippet, description, and then a code block with an example usage. We should be displaying that for this query. It looks like we're instead showing the names/descriptions of other `React.PropTypes.*` methods?

Please investigate and fix!

File: lib/fathead/reactjs/parse.py [hunk 3]
```diff
@@ -110,7 +117,10 @@ def parse_content(self, api):
                 continue
             else:
                 abstract+=(str(tag))
-        return abstract.strip('\n')
+        abstract = self.remove_anchor(abstract)
+        abstract+='\n'
```
Comment (dip-kush) [inline lib/fathead/reactjs/parse.py]: I will do it.

**Comment by dip-kush** [comment]
@moollaza  I will try to fix these issues in a new PR.
